django/db/models/sql/query.py

```diff
@@ -1137,7 +1137,29 @@ def build_filter(self, filter_expr, branch_negated=False, current_negated=False,
 
         # Work out the lookup type and remove it from the end of 'parts',
         # if necessary.
-        value, lookups, used_joins = self.prepare_lookup_value(value, lookups, can_reuse, allow_joins)
+        if isinstance(value, (list, tuple)) and value:
+            klass = type(value)
+            new_value = []
+            new_lookups = []
+            new_used_joins = []
+            # Look up using the iterable of arguments first, in case it's an empty list
+            for inner_value in value:
+                inner_value, lookups, used_joins = self.prepare_lookup_value(inner_value,
+                    lookups,
+                    can_reuse,
+                    allow_joins)
+                new_value.append(inner_value)
+                for lookup in lookups:
+                    if lookup not in new_lookups:
+                        new_lookups.append(lookup)
+                for used_join in used_joins:
+                    if used_join not in new_used_joins:
+                        new_used_joins.append(used_join)
```

Comment:
I think what Anssi is asking you to do is prove that the following produces LEFT OUTER joins rather than INNER JOINs.

```
class Timestamp(models.Model):
    at_time = models.DateTimeField()

class Event(models.Model):
    start = models.ForeignKey(Timestamp, null=True)
    end = models.ForeignKey(Timestamp, null=True)
    actual = models.DateTimeField()

start_datetime = datetime.datetime(2016, 6, 2)
end_datetime = datetime.datetime(2016, 6, 4)
actual = datetime.datetime(2016, 6, 3)
t1 = Timestamp.objects.create(at_time=start_datetime)
t2 = Timestamp.objects.create(at_time=end_datetime)

Event.objects.create(start=t1, end=t2, actual=actual)
Event.objects.create(start=t1, end=None, actual=actual)
Event.objects.create(start=None, end=t2, actual=actual)
Event.objects.create(start=None, end=None, actual=actual)

# this should produce LEFT OUTER JOINS, not INNER JOIN
Event.objects.filter(actual__range=[F('start__at_time'), F('end__at_time')])

```

I'm not sure what those results will be, but the joins should be LEFT OUTER.